### PR TITLE
Upsert small segment merger task in minions

### PR DESCRIPTION
**Pull request by tibrewalpratik17**
PR related to the PEP request: #14305 

Here, we are adding a new minion task to merge small segments in an upsert table. More implementation details in the design doc of the linked issue.

Test plan: Enabled this in one of infinite retention tables in Uber. The tables had ~35k segments initially and after enabling this task for ~2 days we were able to reach ~2k segments. The curve also flattens post reaching ~2k segments. We are using the default configs of this task and the table is generating ~500 segments daily. See attached screenshot.

Screenshot 2024-11-21 at 4 28 22 PM

Few details:
- Unlike upsert compaction, we are enforcing enableSnapshot to be enabled for running this task. This makes sense from correctness perspective.
- The new segment name formed example: compactmerged__table__5__1731976847331__0
- The creation time of the new segment in ZK = max creation time in ZK of merged segments. Note the segment name doesn't have the index creation time value but the task running time. 
- The code merges LLC segment and UploadedRealtime segments which is desired. 
- Right now, we process one merging subtask per partition per task run. We will explore later of allowing multiple merging subtasks per partition per task run.

### CODE REVIEW

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+        // this can be later removed if we want to just do single-segment compaction from this task
+        if (groups.get(0).size() <= 1) {
+          continue;
+        }
+        // TODO see if multiple groups of same partition can be added
+        Map<String, String> configs = new HashMap<>(getBaseTaskConfigs(tableConfig,
+            groups.get(0).stream()
+                .map(x -> x.getSegmentZKMetadata().getSegmentName()).collect(Collectors.toList())));
+        configs.put(MinionConstants.DOWNLOAD_URL_KEY,
+            StringUtils.join(groups.get(0).stream().map(x -> x.getSegmentZKMetadata().getDownloadUrl())
```
Comment (ankitsultana) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java:210]: Let's move the StringUtils stuff to a separate function? Not only improves readability but you can also write a quick UT. Same for `ORIGINAL_SEGMENT_CRC_KEY`

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+          segmentsEligibleForCompactMerge.computeIfAbsent(partitionID, k -> new ArrayList<>())
+              .add(new SegmentMergerMetadata(segment, totalValidDocs, totalInvalidDocs));
+        }
+        break;
+      }
+    }
+
+    segmentsEligibleForCompactMerge.forEach((partitionID, segmentList) -> segmentList.sort((o1, o2) -> {
+      // Sort primarily by creationTime in ascending order
+      if (o1.getSegmentZKMetadata().getCreationTime() < o2.getSegmentZKMetadata().getCreationTime()) {
```
Comment (ankitsultana) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java:274]: We can use Comparator.comparingLong here?

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+
+    // Iterate over each partition and process its segments list
+    for (Map.Entry<Integer, List<SegmentMergerMetadata>> entry : segmentsEligibleForCompactMerge.entrySet()) {
+      int partitionID = entry.getKey();
+      List<SegmentMergerMetadata> segments = entry.getValue();
+      // task config thresholds
+      long validDocsThreshold = Long.parseLong(taskConfigs.getOrDefault(MinionConstants
+              .UpsertCompactMergeTask.MAX_NUM_RECORDS_PER_SEGMENT_KEY,
+          String.valueOf(MinionConstants.UpsertCompactMergeTask.DEFAULT_MAX_NUM_RECORDS_PER_SEGMENT)));
+      long maxRecordsPerTask = Long.parseLong(taskConfigs.getOrDefault(MinionConstants
```
Comment (ankitsultana) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java:293]: Let's leave a TODO for finalizing the semantics. Specifically, we should rely on segment size as our primary guide instead of max rows per task. We could do it in a separate PR which can help drive a discussion too.

File: pinot-core/src/main/java/org/apache/pinot/core/common/MinionConstants.java [hunk 4]
```diff
@@ -199,4 +200,59 @@ public static class UpsertCompactionTask {
      */
     public static final String NUM_SEGMENTS_BATCH_PER_SERVER_REQUEST = "numSegmentsBatchPerServerRequest";
   }
+
+  public static class UpsertCompactMergeTask {
```
Comment (ankitsultana) [inline pinot-core/src/main/java/org/apache/pinot/core/common/MinionConstants.java:204]: broad question: say we run both merge compaction and the segment refresh task for a table. How do we ensure that there aren't any race conditions, because it could be that the two tasks end up processing the same set of segments concurrently.

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskExecutor.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+        throw new IllegalStateException(String.format("Partition id mismatched for %s, expected partition id: %d",
+            segmentName, partitionId));
+      }
+      partitionId = segmentPartitionId;
+      maxCreationTimeOfMergingSegments = Math.max(maxCreationTimeOfMergingSegments,
+          segmentMetadata.getIndexCreationTime());
+
+      String crcFromDeepStorageSegment = segmentMetadata.getCrc();
+      if (!originalSegmentCrcFromTaskGenerator.get(i).equals(crcFromDeepStorageSegment)) {
+        String message = String.format("Crc mismatched between ZK and deepstore copy of segment: %s. Expected crc "
```
[the remark this answers is not in the document]
Comment (tibrewalpratik17) [reply pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskExecutor.java:106]: Yes one is the deepstore upload retry task where it can upload from any replica and cause CRC mismatch. One more might be an edge case during segment commit time maybe (I have not deep-dived into this). We have an issue to track this - https://github.com/apache/pinot/issues/13491

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+
+
+@TaskGenerator
+public class UpsertCompactMergeTaskGenerator extends BaseTaskGenerator {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(UpsertCompactMergeTaskGenerator.class);
+  private static final String DEFAULT_BUFFER_PERIOD = "2d";
+  private static final int DEFAULT_NUM_SEGMENTS_BATCH_PER_SERVER_REQUEST = 500;
+
+  public static class SegmentMergerMetadata {
```
Comment (klsince) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java:62]: private class or `VisibleForTests`?

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+    }
+
+    public long getInvalidDocIds() {
+      return _invalidDocIds;
+    }
+  }
+
+  public static class SegmentSelectionResult {
+
+    private final Map<Integer, List<List<SegmentMergerMetadata>>> _segmentsForCompactMerge;
```
Comment (klsince) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java:88]: `_segmentsForCompactMergeByPartition`? or comment a bit that the key is partitionId

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskExecutor.java [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+      for (RecordReader recordReader : recordReaders) {
+        recordReader.close();
+      }
+    }
+
+    long endMillis = System.currentTimeMillis();
+    LOGGER.info("Finished task: {} with configs: {}. Total time: {}ms", taskType, configs, (endMillis - startMillis));
+
+    List<SegmentConversionResult> results = new ArrayList<>();
+    for (File outputSegmentDir : outputSegmentDirs) {
```
Comment (klsince) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskExecutor.java:151]: just to be sure, could one task generate multiple new segments or just one? If multiple segments could be created, then how to handle cases when some segments failed to get uploaded.

Comment (tibrewalpratik17) [reply pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskExecutor.java:151]: Current implementation ensures generating one segment. But if you think about it even if we generate multiple segments, we don't need to think about rollback much. It's just that whatever segments got uploaded, validDocIds will be uploaded for those and the corresponding segments would be discarded as their docs will be invalid (assuming takeSnapshot has run once).

Comment (tibrewalpratik17) [reply pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskExecutor.java:156]: > For M:N merging, I think the computation of alreadyMergedSegments might need some extensions, as you may want to retry some of the input segments, instead of skipping them if they are in some segment's custom map.

Yeah or we ensure always selecting complete 'x' segments while creating N segments such that sum over x = M. For the sake of simplicity, we are resolving this in task-generation itself currently ensuring to generate multiple X:1 subtasks instead.

File: pinot-core/src/main/java/org/apache/pinot/core/common/MinionConstants.java (hunk 4 above)
Comment (ankitsultana) [reply pinot-core/src/main/java/org/apache/pinot/core/common/MinionConstants.java:228]: Correctness wise we should be fine I guess, but this will still impede or impact the Segment Refresh task. Can you create a separate issue for this with some context on the problem so we could pick it up later?

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskExecutor.java [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+      }
+    }).collect(Collectors.toList());
+
+    // validate if partitionID is same for all small segments. Get partition id value for new segment.
+    int partitionID = getCommonPartitionIDForSegments(segmentMetadataList);
+
+    // get the max creation time of the small segments. This will be the index creation time for the new segment.
+    Optional<Long> maxCreationTimeOfMergingSegments = segmentMetadataList.stream().map(
+        SegmentMetadataImpl::getIndexCreationTime).reduce(Long::max);
+    if (maxCreationTimeOfMergingSegments.isEmpty()) {
```
Comment (ankitsultana) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskExecutor.java:102]: Just to confirm: `numInputSegments` is always greater than 0 right?

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+
+    // Map to store the result: each key (partition) will have a list of groups
+    Map<Integer, List<List<SegmentMergerMetadata>>> groupedSegments = new HashMap<>();
+
+    // Iterate over each partition and process its segments list
+    for (Map.Entry<Integer, List<SegmentMergerMetadata>> entry : segmentsEligibleForCompactMerge.entrySet()) {
+      int partitionID = entry.getKey();
+      List<SegmentMergerMetadata> segments = entry.getValue();
+      // task config thresholds
+      // TODO add output segment size as one of the thresholds
```
Comment (ankitsultana) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java:282]: Can we create a issue for this for tracking?

**Review by ankitsultana — APPROVED** [review]

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskExecutor.java (hunk 9 above)
Comment (tibrewalpratik17) [reply pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskExecutor.java:102]: Yes we have added the check in generator code to ensure `numInputSegments > 1`

File: pinot-core/src/main/java/org/apache/pinot/core/common/MinionConstants.java (hunk 4 above)
Comment (tibrewalpratik17) [reply pinot-core/src/main/java/org/apache/pinot/core/common/MinionConstants.java:228]: https://github.com/apache/pinot/issues/14633

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java (hunk 10 above)
Comment (tibrewalpratik17) [reply pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java:282]: https://github.com/apache/pinot/issues/14634

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java [hunk 11]
```diff
@@ -0,0 +1,424 @@
...
+  public String getTaskType() {
+    return MinionConstants.UpsertCompactMergeTask.TASK_TYPE;
+  }
+
+  @Override
+  public List<PinotTaskConfig> generateTasks(List<TableConfig> tableConfigs) {
+    String taskType = MinionConstants.UpsertCompactMergeTask.TASK_TYPE;
+    List<PinotTaskConfig> pinotTaskConfigs = new ArrayList<>();
+    for (TableConfig tableConfig : tableConfigs) {
+
```
Comment (klsince) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java:117]: nit: move the for-loop logic to a helper method to be more readable

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java [hunk 12]
```diff
@@ -0,0 +1,424 @@
...
+        numTasks++;
+      }
+      LOGGER.info("Finished generating {} tasks configs for table: {}", numTasks, tableNameWithType);
+    }
+    return pinotTaskConfigs;
+  }
+
+  @VisibleForTesting
+  public static SegmentSelectionResult processValidDocIdsMetadata(
+      Map<String, String> taskConfigs,
```
Comment (klsince) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java:223]: format?

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java [hunk 13]
```diff
@@ -0,0 +1,424 @@
...
+      Map<String, SegmentZKMetadata> candidateSegmentsMap,
+      Map<String, List<ValidDocIdsMetadataInfo>> validDocIdsMetadataInfoMap,
+      Set<String> alreadyMergedSegments) {
+    Map<Integer, List<SegmentMergerMetadata>> segmentsEligibleForCompactMerge = new HashMap<>();
+    List<String> segmentsForDeletion = new ArrayList<>();
+    for (String segmentName : validDocIdsMetadataInfoMap.keySet()) {
+      // check if segment is part of completed segments
+      if (!candidateSegmentsMap.containsKey(segmentName)) {
+        LOGGER.warn("Segment {} is not found in the completed segments list, skipping it for {}", segmentName,
+            MinionConstants.UpsertCompactMergeTask.TASK_TYPE);
```
Comment (klsince) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java:233]: I assume we don't need to log task type as the LOGGER prints class name anyway

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java [hunk 14]
```diff
@@ -0,0 +1,424 @@
...
+        if (segment.getCrc() != Long.parseLong(validDocIdsMetadata.getSegmentCrc())) {
+          LOGGER.warn("CRC mismatch for segment: {}, (segmentZKMetadata={}, validDocIdsMetadata={})", segmentName,
+              segment.getCrc(), validDocIdsMetadata.getSegmentCrc());
+          continue;
+        }
+
+        // segments eligible for deletion with no valid records
+        long totalDocs = validDocIdsMetadata.getTotalDocs();
+        if (totalInvalidDocs == totalDocs) {
+          segmentsForDeletion.add(segmentName);
```
Comment (klsince) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java:251]: `break;` early here? 

 some of the `else keyword` can be saved with those `continued;`

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java [hunk 15]
```diff
@@ -0,0 +1,424 @@
...
+        }
+      }
+      // Add the last group
+      if (!currentGroup.isEmpty()) {
+        groups.add(new ArrayList<>(currentGroup)); // Add a copy of the current group
+      }
+
+      // Sort groups by total invalidDocs in descending order, if invalidDocs count are same, prefer group with
+      // higher number of small segments in them
+      // remove the groups having only 1 segments in them -- this check can be later removed if we want single-segment
```
Comment (klsince) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java:332]: `--` meant a TODO here?

**Review by klsince — APPROVED** [review]
LGTM

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java (hunk 11 above)
Comment (tibrewalpratik17) [reply pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java:117]: yes agreed! let me do this in a follow-up.
